civis/parallel.py

```diff
@@ -0,0 +1,719 @@
+        self._future.remote_func_output = None  # `get` reads results from here
+        self._future.result_fetched = False  # Did we get the result?
+        self._future.add_done_callback(
+            self._make_fetch_callback(self._callback, self._client))
+
+    @staticmethod
+    def _make_fetch_callback(joblib_callback, client):
+        """Create a closure for use as a callback on the ContainerFuture"""
+        def _fetch_result(fut):
+            """Retrieve outputs from the remote function.
+            Run the joblib callback only if there were no errors.
+
+            Parameters
+            ----------
+            fut : :class:`~civis.futures.ContainerFuture`
+                A Future which represents a Civis job. Created by a
+                :class:`~_ContainerShellExecutor`.
+
+            Note
+            ----
+            We can't return data from a callback, so the remote
+            function output is attached to the Future object
+            as a new attribute ``remote_func_output``.
+            """
+            if fut.succeeded():
+                log.debug(
+                    "Ran job through Civis. script ID: %d, run ID: %d;"
+                    " job succeeded!", fut.job_id, fut.run_id)
+            elif fut.cancelled():
+                log.debug(
+                    "Ran job through Civis. script ID: %d, run ID: %d;"
+                    " job cancelled!", fut.job_id, fut.run_id)
+            else:
+                log.error(
+                    "Ran job through Civis. script ID: %d, run ID: %d;"
+                    " job failure!", fut.job_id, fut.run_id)
+
+            try:
+                # Find the output file ID from the run outputs.
+                run_outputs = client.scripts.list_containers_runs_outputs(
+                    fut.job_id, fut.run_id)
+                if run_outputs:
+                    output_file_id = run_outputs[0]['object_id']
+                    res = _robust_result_download(output_file_id, client,
+                                                  n_retries=5, delay=1.0)
+                    fut.remote_func_output = res
+                    log.debug("Downloaded and deserialized the result.")
+            except BaseException as exc:
+                # If something went wrong when fetching outputs, record the
+                # exception so we can re-raise it in the parent process.
+                # Catch BaseException so we can also re-raise a
+                # KeyboardInterrupt where it can be properly handled.
+                log.debug('Exception during result download: %s', str(exc))
+                fut.remote_func_output = exc
+            else:
+                fut.result_fetched = True
+                if not fut.cancelled() and not fut.exception():
+                    # The next job will start when this callback is called.
+                    # Only run it if the job was a success.
+                    joblib_callback(fut.remote_func_output)
```

Comment:
We should support an `archive_on_success` option (maybe as opposed to `hidden`?). The idea is that all jobs would be visible in the API and the website (so `hidden=False`), but if they complete successfully, they get archived with a call to `put_containers_archive`. That call would go here.

This way people can see progress and look at running jobs, but the jobs clean themselves up unless they throw an error, in which case they are visible for inspection/debugging.